library/ord-map.lisp

```diff
@@ -0,0 +1,204 @@
+(coalton-library/utils:defstdlib-package :coalton-library/ord-map
+    (:use
+     :coalton
+     :coalton-library/classes
+     :coalton-library/tuple
+     :coalton-library/functions)
+  (:local-nicknames (#:tree :coalton-library/ord-tree)
+                    (#:iter :coalton-library/iterator))
+  (:shadow #:Map #:empty)
+  (:export
+   #:Map
+   #:empty
+   #:lookup
+   #:insert
+   #:replace
+   #:replace-or-insert #:insert-or-replace
+   #:remove
+   #:keys
+   #:values
+   #:entries
+   #:collect!
+   #:update
+   #:merge))
+(cl:in-package :coalton-library/ord-map)
+
+(coalton-toplevel
+  (define-type (MapPair :key :value)
+    (MapPair :key :value)
+
+    ;; hack: when searching for an entry, we need a thing that holds just a key, but is of type `MapPair :key
+    ;; :any', to pass to `tree:lookup' or `tree:remove'. A `Map' will never hold a `JustKey'.
+    (JustKey :key))
+
+  (define (key pair)
+    (match pair
+      ((MapPair k _) k)
+      ((JustKey k) k)))
+
+  (define (value pair)
+    (match pair
+      ((MapPair _ v) v)
+      (_ (error "Misuse of `JustKey': Cannot read the Value of a JustKey"))))
+
+  (define-instance (Into (MapPair :key :value) (Tuple :key :value))
+    (define (into pair)
+      (Tuple (key pair) (value pair))))
+
+  (define-instance ((Eq :key) => (Eq (MapPair :key :value)))
+    (define (== left right)
+      (== (key left) (key right))))
+
+  (define-instance ((Ord :key) => (Ord (MapPair :key :value)))
+    (define (<=> left right)
+      (<=> (key left) (key right))))
+
+  (repr :transparent)
+  (define-type (Map :key :value)
+    "A red-black binary tree which associates each :KEY with a :VALUE, sorted by `<=>' on the keys and unique by `==' on the keys."
+    (%Map (tree:Tree (MapPair :key :value))))
+
+  (declare empty (Map :key :value))
+  (define empty
+    "A Map containing no mappings."
+    (%Map tree:Empty))
+
+  (declare lookup ((Ord :key) => ((Map :key :value) -> :key -> (Optional :value))))
+  (define (lookup mp k)
+    "Retrieve the value associated with K in MP, or None if MP does not contain K."
+    (match mp
+      ((%Map tre) (coalton-library/classes:map value (tree:lookup tre (JustKey k))))))
+
+  (declare insert ((Ord :key) => ((Map :key :value) -> :key -> :value -> (Optional (Map :key :value)))))
+  (define (insert mp k v)
+    "Associate K with V in MP. If MP already contains a mapping for K, return None."
+    (coalton-library/classes:map %Map
```

Comment:
`coalton-library/classes` is already used

Reply:
`Map` is shadowed

Reply:
ah